### PR TITLE
Create Generic Launcher and add launcher for rancher-compose-executor

### PR DESCRIPTION
**Pull request by wlan0**
@ibuildthecloud @cjellick @alena1108

### CODE REVIEW

**Comment by wlan0** [comment]
Now that we are going to launch multiple externalhandlers (go-machine-service, and rancher-compose-executor), I have moved the common code for launching externalhandlers into an abstract class. 

I have also added a launcher for launching rancher-compose-executor in the same PR.

**Comment by wlan0** [comment]
Please do not merge this PR yet. This can be merged after this https://github.com/rancher/cattle/pull/808 is merged

**Comment by wlan0** [comment]
The test failure is random/unrelated

File: code/implementation/docker/machine/src/main/java/io/cattle/platform/docker/machine/launch/ComposeExecutorLauncher.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    protected void setEnvironment(Map<String, String> env) {
+        Credential cred = getCredential();
+        env.put("RANCHER_ACCESS_KEY", cred.getPublicValue());
+        env.put("RANCHER_SECRET_KEY", cred.getSecretValue());
+        env.put("RANCHER_URL", ServerContext.getHostApiBaseUrl(BaseProtocol.HTTP));      
+    }
+
+    @Override
+    protected LockDefinition getLock() {
+        return LauncherLockDefinitions.MachineLauncherLock();
```
Comment (ibuildthecloud) [inline code/implementation/docker/machine/src/main/java/io/cattle/platform/docker/machine/launch/ComposeExecutorLauncher.java]: I think you want the compose lock

File: code/implementation/docker/machine/src/main/java/io/cattle/platform/docker/machine/launch/ComposeExecutorLauncher.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    protected String binaryPath() {
+        return COMPOSE_EXECUTOR_BINARY.get();
+    }
+
+    @Override
+    protected void setEnvironment(Map<String, String> env) {
+        Credential cred = getCredential();
+        env.put("RANCHER_ACCESS_KEY", cred.getPublicValue());
+        env.put("RANCHER_SECRET_KEY", cred.getSecretValue());
+        env.put("RANCHER_URL", ServerContext.getHostApiBaseUrl(BaseProtocol.HTTP));
```
Comment (ibuildthecloud) [inline code/implementation/docker/machine/src/main/java/io/cattle/platform/docker/machine/launch/ComposeExecutorLauncher.java]: Seems like this should be common.  I'd prefer your for executor use `CATTLE_*` vars to be consisten with the rest of the endpoints.

File: code/implementation/docker/machine/src/main/java/io/cattle/platform/docker/machine/launch/LauncherLockDefinitions.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package io.cattle.platform.docker.machine.launch;
+
+import io.cattle.platform.lock.definition.AbstractLockDefinition;
+import io.cattle.platform.lock.definition.LockDefinition;
+
+public class LauncherLockDefinitions extends AbstractLockDefinition implements LockDefinition {
+
+    public LauncherLockDefinitions(String lockId) {
```
Comment (ibuildthecloud) [inline code/implementation/docker/machine/src/main/java/io/cattle/platform/docker/machine/launch/LauncherLockDefinitions.java]: Make private

**Pull request merged**